### PR TITLE
Proposal for request results

### PR DESCRIPTION
**Pull request by irsdkv**
Related to #2

### CODE REVIEW

**Comment by irsdkv** [comment]
Proposal for request results #3

File: src/generated/info/mod.rs [hunk 1]
```diff
@@ -34,71 +33,32 @@ impl From<&info::Version> for Version {
...
+                info::InfoResult_Result::INFORMATION_NOT_RECEIVED_YET => Err(
+                    super::super::RequestError::MavErr(InfoError::InformationNotReceivedYet(
+                        verison_response.get_info_result().get_result_str().into(),
+                    )),
+                ),
+                info::InfoResult_Result::SUCCESS => {
+                    Ok(Version::from(verison_response.get_version()))
+                }
+            },
+            Err(e) => Err(super::super::RequestError::RpcErr(e)),
```
Comment (irsdkv) [inline src/generated/info/mod.rs]: So there is new draft to generate template to [this type ](https://github.com/mavlink/MAVSDK-Rust/pull/4/files#diff-b4aea3e418ccdb71239b96952d9cddb6R14-R16)

Comment (irsdkv) [inline src/generated/info/mod.rs]: 1

File: src/lib.rs [hunk 2]
```diff
@@ -11,18 +11,17 @@ pub struct System {
...
-#[derive(Debug)]
-pub enum MavsdkResult<T, ErrT> {
-    Ok(T),
-    Err(MavsdkError<ErrT>),
-    RpcErr(grpcio::Error),
+trait FromRpcResult<T> {
+    fn from_rpc_result(rpc_result: ::grpcio::Result<T>) -> Self;
 }
 
+pub type RequestResult<SuccessType, PluginMavErr> = Result<SuccessType, RequestError<PluginMavErr>>;
```
Comment (irsdkv) [inline src/lib.rs:23]: @JonasVautherin what are you think about that?

File: examples/info.rs [hunk 3]
```diff
@@ -25,7 +25,18 @@ fn main() {
...
+        Err(error) => match error {
+            RequestError::MavErr(mav_err) => match mav_err {
+                info::InfoError::Unknown(s) => println!("Unknown MAVLink error ({:?})", s),
+                info::InfoError::InformationNotReceivedYet(s) => {
+                    println!("Information not received yet ({:?})", s)
+                }
+            },
+            RequestError::RpcErr(rpc_err) => println!("RPC error: {:?}", rpc_err),
+        },
+    }
```
Comment (irsdkv) [inline examples/info.rs]: And there is example of use

File: src/generated/info/mod.rs (hunk 1 above)
Comment (irsdkv) [inline src/generated/info/mod.rs]: So there is new draft to generate template to this type

File: examples/info.rs [hunk 4]
```diff
@@ -25,7 +25,18 @@ fn main() {
...
 }
```
Comment (irsdkv) [inline examples/info.rs:42]: And there is example of use

Comment (JonasVautherin) [reply examples/info.rs:42]: Is it possible, in Rust, to write directly something like the following?

```rust
    match get_version_result {
        Ok(version) => println!("Success: {:?}", version),
        Err(RequestError::MavErr(mav_err) => match mav_err => { ... },
        Err(RequestError::RpcErr(rpc_err) => ...
    }
```

Comment (irsdkv) [reply examples/info.rs]: > Is it possible, in Rust?

Unfortunately no.
Matching should go "down the tree".
But this allow to implement good structured error processing.

Comment (irsdkv) [reply examples/info.rs:42]: Yes, it is possible

And also:

```rust
Err(ref error) => {
   if let RequestError::MavErr(mav_err) = error {
        println!("MAVLink Error: {:?}", mav_err)
    }
    if let RequestError::RpcErr(rpc_err) = error {
        println!("RPC error: {:?}", rpc_err)
   }
}

Comment (irsdkv) [reply examples/info.rs]: Or 
```rust
Err(ref error) => {
   if let RequestError::MavErr(mav_err) = error {
        println!("MAVLink Error: {:?}", mav_err)
    }
    if let RequestError::RpcErr(rpc_err) = error {
        println!("RPC error: {:?}", rpc_err)
   }
}
```

Comment (irsdkv) [reply examples/info.rs:42]: So my first answer (deleted now) was wrong. It is possible :)

File: src/generated/info/mod.rs [hunk 5]
```diff
@@ -34,71 +33,32 @@ impl From<&info::Version> for Version {
...
-    fn from(rpc_get_version_response: info::GetVersionResponse) -> Self {
-        Self {
-            info_result: InfoResult::from(rpc_get_version_response.get_info_result()),
-            version: Version::from(rpc_get_version_response.get_version()),
+impl FromRpcResult<info::GetVersionResponse> for GetVersionResult {
+    fn from_rpc_result(
+        rpc_get_version_response: ::grpcio::Result<info::GetVersionResponse>,
+    ) -> Self {
+        match rpc_get_version_response {
+            Ok(verison_response) => match verison_response.get_info_result().get_result() {
```
Comment (JonasVautherin) [inline src/generated/info/mod.rs]:
```suggestion
            Ok(version_response) => match version_response.get_info_result().get_result() {
```

File: src/lib.rs (hunk 2 above)
Comment (JonasVautherin) [reply src/lib.rs:23]: Is that synctactic sugar abstracting away `RequestError`? Fancy :sunglasses:.

**Review by JonasVautherin — APPROVED** [review]
Looks good to me, for what I can understand :laughing:. But I like the example, I think it works well. And pattern matching on the result is super cool for handling errors :tada:!

File: src/lib.rs (hunk 2 above)
Comment (irsdkv) [reply src/lib.rs:23]: Yes, it’s hard not to use such nice things :)

**Comment by irsdkv** [comment]
Okay. I think at this moment we can go with this version. 
Any case, we can add or change something if necessary

**Pull request merged**